### PR TITLE
Strict mode

### LINKED ISSUE #51: Minor findings in installation, startup and usage of EMBArk
**Describe the bug**
During the installation the following errors came:

kép

kép

kép

Starting EMBArk displays these issues:

kép

kép

kép

Uploading a firmware file works fine but the path looks like this:

kép

The progress bar did not go to 100% but the san has been finished:

kép

I was able to reproduce this issue more times. I know the scan is finished when the fan slows down in the computer but the progress bar does not go to 100%.

I entered these parameters for the scan:
kép

but the report contains these:
kép

It would be very useful to show real version numbers and proper data in the report, because if I have more scans difficult to check the real versions if the data is incomplete.

**To Reproduce**
1. EMBArk installation (default mode)
2. Start EMBArk: sudo ./run-server.sh
3. Use the firmware available here:
https://cyberforat.squat.net/openwrt/binary_images/linksys_official/WRT54GS_3.17.4_US_code.zip
5. See error

**Expected behavior**
Please fix the errors if posible.

**Screenshots**
Attached above.

**Desktop (please complete the following information):**
 - OS: Ubuntu 22.04.1 LTS Desktop

### PR DESCRIPTION
**Pull request by BenediktMKuehne**
**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
---
Add strict mode to EMBArk installer

**What is the current behavior?** (You can also link to an open issue here)
---
Bugs 😞 

**What is the new behavior (if this is a feature change)? If possible add a screenshot.**
---
Catch some errors

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
---
⚠️ not breaking, but changes to the git submodule init commmands may require manual intervention.
(moved ownership of emba from root to user. Therefore removing the emba directory and `git reset --hard` might be required
to successfully update)

**Other information**:
---

### CODE REVIEW

File: installer.sh [hunk 1]
```diff
@@ -525,11 +518,11 @@ uninstall (){
   # delete user www-embark and reset visudo
   echo -e "$ORANGE""$BOLD""Delete user""$NC"
   # sed -i 's/www\-embark\ ALL\=\(ALL\)\ NOPASSWD\:\ \/app\/emba\/emba.sh//g' /etc/sudoers #TODO doesnt work yet
-  userdel www-embark
+  userdel www-embark || true
```
Comment (m-1-k-3) [inline installer.sh]: could we do a quick check if the user is on the host? like this pseudo code?
If ! grep www-embark /etc/passwd; then
userdel
fi

File: installer.sh [hunk 2]
```diff
@@ -525,11 +518,11 @@ uninstall (){
...
 
   # delete .env
   echo -e "$ORANGE""$BOLD""Delete env""$NC"
-  rm ./.env
+  rm ./.env || true
```
Comment (m-1-k-3) [inline installer.sh]: please always check if the file is available before doing something with it:
if [[ -f ./.env ]]; then
rm
fi

File: installer.sh [hunk 3]
```diff
@@ -542,13 +535,13 @@ uninstall (){
 
   # delete/uninstall EMBA
   echo -e "$ORANGE""$BOLD""Delete EMBA?""$NC"
-  docker network rm emba_runs
+  docker network rm emba_runs || true
```
Comment (m-1-k-3) [inline installer.sh]: could we check if the network device is available before trying to remove it?

File: installer.sh [hunk 4]
```diff
@@ -542,13 +535,13 @@ uninstall (){
...
   git submodule foreach git reset --hard
-  git submodule deinit --all
+  git submodule deinit --all -f
 
   # stop&reset daemon
-  systemctl stop embark.service
-  systemctl disable embark.service
+  systemctl stop embark.service || true
```
Comment (m-1-k-3) [inline installer.sh]: could we do a check if the service is running?

File: run-server.sh [hunk 5]
```diff
@@ -135,11 +134,10 @@ fi
 if ! [[ -d /var/www/conf/cert ]]; then
   mkdir /var/www/conf/cert
 fi
-cp -u "$PWD"/cert/embark.local /var/www/conf/cert
+cp -u "$PWD"/cert/embark.local.crt /var/www/conf/cert
```
Comment (m-1-k-3) [inline run-server.sh]: please do a check if the files are available

File: installer.sh [hunk 6]
```diff
@@ -14,16 +14,18 @@
 
 # Description: Installer for EMBArk
 
+REFORCE=0
```
Comment (m-1-k-3) [inline installer.sh]: are these exports or locals? Pls define here for clarification.

**Review by m-1-k-3 — CHANGES_REQUESTED** [review]
looks quite good. Please review my comments. AFterwards I will do another installation

**Comment by BenediktMKuehne** [comment]
@m-1-k-3  should be ready for another round

File: helper/helper_embark_general.sh [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+#
+# Author(s): Benedikt Kuehne
+
+# Description: Helper functions
+
+docker_image_rm(){
+  # removes image by name and version
+  # $1 name
+  # $2 version
+  local IMAGE_NAME_="$1"
```
Comment (m-1-k-3) [inline helper/helper_embark_general.sh]: for init an empty var in your local function ${1:-} will do the job:

local IMAGE_NAME_="${1:-}"

File: helper/helper_embark_general.sh [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  local IMAGE_VERSION_="$2"
```
Comment (m-1-k-3) [inline helper/helper_embark_general.sh]: init this var with ${2:-}

File: helper/helper_embark_general.sh [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      done
+    fi
+    echo -e "$GREEN$BOLD""Removing ""$IMAGE_NAME_"":""$IMAGE_VERSION_" "docker image""$NC\\n"
+    docker image rm "$IMAGE_NAME_"":""$IMAGE_VERSION_" -f
+  fi
+}
+
+docker_network_rm(){
+  # removes docker networks by name
+  local NET_ID
```
Comment (m-1-k-3) [inline helper/helper_embark_general.sh]: init this var with
local NET_ID=""

File: helper/helper_embark_general.sh [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  if docker network ls | grep -E "$1"; then
```
Comment (m-1-k-3) [inline helper/helper_embark_general.sh]: please start setting up the vars in the beginning of the func. E.g.
local NAME_OF_VAR="${1:-}"
local NAME_OF_VAR2="${2:-}"

File: helper/helper_embark_general.sh [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    echo -e "\n$GREEN""$BOLD""Found ""$1"" - removing it""$NC"
+    NET_ID=$(docker network ls | grep -E "$1" | awk '{print $1}')
+    echo -e "$GREEN""$BOLD""Remove ""$1"" network""$NC"
+    docker network rm "$NET_ID" 
+  fi
+}
+
+copy_file(){
+  # check and copy file forcing overwrite
+  # $1 : source
```
Comment (m-1-k-3) [inline helper/helper_embark_general.sh]: please init the vars here
local SOURCE="${1:-}"
local DEST="${2:-}"

File: helper/helper_embark_general.sh [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    return 1
+  elif  ! [[ -d $(dirname "$2") ]] ; then
+    echo -e "\\n$RED""Could not find ""$2""$NC\\n"
+    return 1
+  fi
+  cp -f "$1" "$2"
+}
+
+enable_strict_mode() {
+  local STRICT_MODE_="$1"
```
Comment (m-1-k-3) [inline helper/helper_embark_general.sh]: see here https://github.com/m-1-k-3/emba/blob/5a0a58b76b81ba42d75bc4efbadb6589e0a886b1/helpers/helpers_emba_helpers.sh#L185

File: installer.sh [hunk 13]
```diff
@@ -46,68 +44,75 @@ print_help() {
...
 # Source: https://stackoverflow.com/questions/4023830/how-to-compare-two-strings-in-dot-separated-version-format-in-bash
 version() { echo "$@" | awk -F. '{ printf("%d%03d%03d%03d\n", $1,$2,$3,$4); }'; }
 
 write_env() {
-  # TODO change to locals
+  local SUPER_PW="embark"
+  local SUPER_EMAIL="idk@lol.com"
+  local SUPER_USER="superuser"
+
+  local RANDOM_PW
```
Comment (m-1-k-3) [inline installer.sh]: Please init all vars:
local RANDOM_PW=""

File: installer.sh [hunk 14]
```diff
@@ -46,68 +44,75 @@ print_help() {
...
+  local DJANGO_SECRET_KEY
```
Comment (m-1-k-3) [inline installer.sh]: init
local DJANGO_SECRET_KEY=""

File: installer.sh [hunk 15]
```diff
@@ -138,86 +143,30 @@ dns_resolve(){
...
-    echo -e "\n$GREEN""$BOLD""Found EMBArk_backend network - removing it""$NC"
-    NET_ID=$(docker network ls | grep -E "embark_backend" | awk '{print $1}')
-    echo -e "$GREEN""$BOLD""Remove EMBArk_backend network""$NC"
-    docker network rm "$NET_ID" 
-  fi
-  
 }
 
 install_debs() {
+  local DOCKER_COMP_VER
```
Comment (m-1-k-3) [inline installer.sh]: =""

**Comment by m-1-k-3** [comment]
a few late notes for this PR. Will do a fresh install during the weekend ...

**Review by m-1-k-3 — APPROVED** [review]

**Pull request merged**